eth/db/chain.py

```diff
@@ -93,15 +94,18 @@ def get_block_uncles(self, uncles_hash: Hash32) -> Tuple[BlockHeaderAPI, ...]:
     def _set_as_canonical_chain_head(
             cls,
             db: DatabaseAPI,
-            block_hash: Hash32,
+            header: BlockHeaderAPI,
             genesis_parent_hash: Hash32,
     ) -> Tuple[Tuple[BlockHeaderAPI, ...], Tuple[BlockHeaderAPI, ...]]:
+        # performance optimization, short circuit if we're just adding a single header
         try:
-            header = cls._get_block_header_by_hash(db, block_hash)
-        except HeaderNotFound:
-            raise ValueError(
-                f"Cannot use unknown block hash as canonical head: {header.hash}"
-            )
+            current_canonical_head = cls._get_canonical_head_hash(db)
+        except CanonicalHeadNotFound:
+            current_canonical_head = None
+        if current_canonical_head and header.parent_hash == current_canonical_head:
+            cls._add_block_number_to_hash_lookup(db, header)
+            db.set(SchemaV1.make_canonical_head_hash_lookup_key(), header.hash)
+            return (header,), tuple()
```

Comment:
Cool, doing this optimization makes sense, but the logic duplication (adding block number lookup, and setting the canonical head hash) concerns me. If something updates further down in the method, it would be *really* easy to forget to do it here.

One alternative:

- In the if block, only set `new_canonical_headers, old_canonical_headers = (header,), ()`
- Move the following 3 lines of `{old,new}_canonical_headers` generation into the else block

It does one extra db lookup at the `_get_canonical_block_hash` to check for an old header.

If you think that extra lookup is critical, and worth some more PR thrash, then you could move the `for h in new_canonical_headers` block into a `cls._decanonicalize_old_headers(db, new_headers)` call, inside the else block.

Reply:
Yes, great point!

Reply:
Surprisingly enough the extra lookup added ~9% to import time. I pushed a change which keeps it out but also removes some duplicated code.